src/output.cpp

```diff
@@ -236,163 +236,236 @@ std::string Output::get_helper_error_msg(int func_id, int retcode) const
+    }
+    case Type::cgroup_path: {
+      return bpftrace.resolve_cgroup_path(
+          reinterpret_cast<const AsyncEvent::CgroupPath *>(value.data())
+              ->cgroup_path_id,
+          reinterpret_cast<const AsyncEvent::CgroupPath *>(value.data())
+              ->cgroup_id);
+    }
+    case Type::strerror: {
+      return strerror(read_data<uint64_t>(value.data()));
+    }
+    case Type::pointer:
+    case Type::reference: {
+      std::ostringstream res;
+      res << "0x" << std::hex << read_data<uint64_t>(value.data());
+      return res.str();
+    }
+    default: {
+      assert(type.IsNoneTy());
+      return "";
     }
-    return std::to_string(read_data<uint64_t>(value.data()) / div);
-  } else if (type.IsProbeTy())
-    return bpftrace.resolve_probe(read_data<uint64_t>(value.data()));
-  else if (type.IsTimestampTy())
-    return bpftrace.resolve_timestamp(
-        reinterpret_cast<AsyncEvent::Strftime *>(value.data())->mode,
-        reinterpret_cast<AsyncEvent::Strftime *>(value.data())->strftime_id,
-        reinterpret_cast<AsyncEvent::Strftime *>(value.data())->nsecs);
-  else if (type.IsMacAddressTy())
-    return bpftrace.resolve_mac_address(value.data());
-  else if (type.IsCgroupPathTy())
-    return bpftrace.resolve_cgroup_path(
-        reinterpret_cast<AsyncEvent::CgroupPath *>(value.data())
-            ->cgroup_path_id,
-        reinterpret_cast<AsyncEvent::CgroupPath *>(value.data())->cgroup_id);
-  else if (type.IsStrerrorTy())
-    return strerror(read_data<uint64_t>(value.data()));
-  else if (type.IsPtrTy() || type.IsRefTy()) {
-    std::ostringstream res;
-    res << "0x" << std::hex << read_data<uint64_t>(value.data());
-    return res.str();
-  } else {
-    assert(type.IsNoneTy());
-    return "";
   }
 }
 
+std::string Output::map_key_str(BPFtrace &bpftrace,
+                                const SizedType &arg,
+                                const std::vector<uint8_t> &data) const
+{
+  std::ostringstream ptr;
+  switch (arg.GetTy()) {
+    case Type::integer:
+    case Type::kstack:
+    case Type::ustack:
+    case Type::timestamp:
+    case Type::ksym:
+    case Type::usym:
+    case Type::inet:
+    case Type::username:
+    case Type::probe:
+    case Type::string:
+    case Type::buffer:
+    case Type::pointer:
+    case Type::array:
+    case Type::mac_address:
+    case Type::record:
+      return value_to_str(bpftrace, arg, data, false, 1, true);
+    case Type::tuple: {
+      std::vector<std::string> elems;
+      for (auto &field : arg.GetFields()) {
+        std::vector<uint8_t> elem_data(data.begin() + field.offset,
+                                       data.begin() + field.offset +
+                                           field.type.GetSize());
+        elems.push_back(
+            value_to_str(bpftrace, field.type, elem_data, false, 1, true));
+      }
+      return tuple_to_str(elems, true);
+    }
+    default:
```

Comment:
Instead of default, list out all the other types? Compiler will warn if new type is added